containerized code only for remote mpi mapping hpc type (#5667)
aiidateam/aiida-core: docs/source/topics/data_types.rst

        filepath_executable='executable.exe'
    )

.. _topics:data_types:core:code:containerized:

ContainerizedCode
-----------------

.. versionadded:: 2.1

The :class:`~aiida.orm.nodes.data.code.containerized.ContainerizedCode` class allows running an executable within a container image on a target computer.
The data plugin stores the following information in the database:

> * ``image_name``: The name of the container image (e.g., ``docker://alpine:3``).

Reviewer: @unkcpz can you confirm this is correct? If so, I will correct it.

> * ``image_name``: The name of the container image (e.g., ``docker://alpine:3``).

Author: Yes, it is correct, can be the abs path to `.sif` image file download by `singularity pull`

* ``filepath_executable``: The filepath of the executable within the container (e.g. ``/usr/bin/bash``).
* ``engine_command``: The bash command to invoke the container image (e.g. ``singularity exec --bind $PWD:$PWD {image_name}``).
  The exact form of this command will depend on the containerization technology that is used.
* ``computer``: The :class:`~aiida.orm.computers.Computer` on which to run the container.

.. important::

    If the ``engine_command`` contains variables (such as in the ``singularity exec --bind $PWD:$PWD {image_name}`` example), it is crucial that the ``Computer`` needs to have the ``use_double_quotes`` setting set to ``True``.
    By default, a ``Computer`` will use single quotes to escape command line arguments and so the ``$PWD`` would not be expanded.
    If this wasn't defined as such when the ``Computer`` was created, it can be changed through the API:

    .. code-block::

        computer = load_computer('some-computer')
        computer.set_use_double_quotes(True)

> 

Reviewer: @unkcpz any idea what information would be relevant here?

> 

Reviewer 2: What I mean is to give an example of how the command line for an MPI job could look like, which illustrates that the container runtime itself is called through MPI.

> 

Reviewer: I have added the test and added an example of the runline that is generated when MPI is enabled.

> 

Reviewer 2: thanks Seb, I think your explanation is very clear.

.. note::

    The container image can be already download and stored on the target computer or can reference an image on a container index, such as Docker Hub.
    In the latter case, the first time that the containerized code is run, the image will be automatically pulled from the index, which may take a while if the image is large.

.. _topics:data_types:core:code:installed:containerized:setup:

Setup
^^^^^

A ``ContainerizedCode`` can be created through the CLI as well as the API.
The following examples show how to setup running ``bash`` in a base Docker container through Singularity to be run on the ``Computer`` named ``some-computer``:

.. tab-set::

    .. tab-item:: CLI

        .. code-block:: console

            verdi code create core.code.containerized \
                --non-interactive \
                --label containerized-code \
                --default-calc-job-plugin core.arithmetic.add \
                --computer some-computer \
                --filepath-executable "/bin/sh" \
                --image-name "docker://alpine:3" \

>                 --engine-command "singularity exec --bind $PWD:$PWD {image_name}"

Reviewer 2: By the way, it appears to me that the file path in the `engine_command` is not supposed to be absolute, while we require the path to the executable of the `Code` to be absolute.

This is fine with me, just wanted to point it out.
I guess one can argue that the result of the calculation is going to be less dependent on the exact version of the container runtime than on the exact version of the simulation code.
 
P.S. There have been several instances in the past where I've found the requirement of an absolute file path to be onerous (e.g. I rebuilt packages with spack, the `module load` command would have remained the same but the absolute path changed because of some hash, so I had to re-setup all the codes). It is not _bad_ to ask for the absolute path of the Code, it's just a bit pedantic not to give the user the choice. 

P.P.S. Inside a container, the requirement of an absolute file path for the code executable arguably loses its original purpose since the provenance on the code version is carried by the image name. To keep things simple, we can keep the requirement nevertheless.

Reviewer: > By the way, it appears to me that the file path in the engine_command is not supposed to be absolute, while we require the path to the executable of the Code to be absolute.

I actually already addressed this in the latest change. The `ContainerizedCode` overrides the `filepath_executable` property and no longer requires that it is absolute.

Reviewer 2: Ah, cool!

    .. tab-item:: API

        .. code-block:: python

            from aiida.orm import ContainerizedCode, load_computer

            code = ContainerizedCode(
                computer=load_computer('some-computer')
                filepath_executable='/bin/sh'
                image_name='docker://alpine:3',
                engine_command='singularity exec --bind $PWD:$PWD {image_name}'
            ).store()

.. _topics:data_types:core:code:installed:containerized:support:

Supported container technologies
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The ``ContainerizedCode`` is compatible with a variety of containerization technologies:

Reviewer 2: In order to make this subsection most useful, please add for each engine an example of what to put into the `engine_command` field.

Reviewer: I agree, and I already put this in [my comment](https://github.com/aiidateam/aiida-core/pull/5667#issuecomment-1282274083). I didn't add these examples since they weren't there in the original PR and I have no idea what they should be. I guess if @unkcpz already has them, we can quickly add them. But if that requires more work, maybe this can be done at a later stage.

Author: I add the command to line below.

* `Singularity <https://singularity-docs.readthedocs.io/en/latest/>`__
* `Sarus <https://sarus.readthedocs.io/en/stable/>`__

> * `NVIDIA/enroot <https://github.com/NVIDIA/enroot/>`__

Reviewer 2: on second thought, vanilla `enroot` may fall into the `docker` category after all (not supported), since you would need to call `mpirun` _inside_  the container...

are there any plans to support such scenarios?

P.S. It actually already possible to use enroot containers with AiiDA via the pyxis plugin for slurm, which adds some extra slurm pragmas like
```
#SBATCH --container-image ...
#SBATCH --container-mounts
```
One can then tell these to AiiDa as extra scheduler parameters.
It's not great from a provenance perspective, though, since the mounts, image, etc. are all recorded in a slurm-specific way.

Reviewer 2: Sure

Reviewer: I removed the mention of enroot for now until we figure out if and how it can be supported.

* `Shifter <https://github.com/NERSC/shifter>`__

> Using `Docker <https://www.docker.com/>`__ directly is currently not supported, since Docker has to run with root permissions and it cannot be launched as a normal MPI program to propagate execution context to the container application.

Reviewer 2: I get the MPI part, but I don't understand what the permissions for running docker has to do with AiiDA compatibility?

Either the user has the rights to run docker or they don't - I don't think this is relevant for the integration.

```suggestion
Using the `Docker <https://www.docker.com/>`__ container runtime is currently not supported, since Docker cannot be launched as a MPI program to propagate execution context to the container application.
```

Reviewer: @unkcpz fine to remove the mention of the root permissions?

Author: > I get the MPI part, but I don't understand what the permissions for running docker has to do with AiiDA compatibility?

This is for the file created inside the container and file generated inside the container. If in the `Dockerfile` no `USER` is specified, the container with run with the `UID=0` the root user. The file it created can not be cleaned up by clean workdir for example.

Reviewer: For the time being, I have rephrased it as

```

Using `Docker <https://www.docker.com/>`__ directly is currently not supported because:

* Docker needs to be run as the root user and the files created in the working directory inside the container will be owned by root, which prevents AiiDA from deleting those files after execution.
* Docker cannot be launched as a normal MPI program to propagate execution context to the container application.

Support may be added at a later time.
```

is that correct and intelligible?

Author: That is the default case I'd say but there is workaround and not so exact. Maybe better use the phrase from their official documentation is better?

["The Docker daemon always runs as the root user."](https://docs.docker.com/engine/install/linux-postinstall/#:~:text=The%20Docker%20daemon%20always%20runs,members%20of%20the%20docker%20group.)

> The Docker daemon always runs as the root user and the files created in the working directory inside the container will usually be owned by root if no `uid` is not specified in image, which prevents AiiDA from deleting those files after execution.
